core/src/main/java/software/amazon/awssdk/interceptor/ExecutionInterceptor.java

```diff
@@ -0,0 +1,297 @@
+ *     the order than interceptors later in the file.</li>
+ *
+ *     <li><i>Service Interceptors</i>. Interceptors earlier in the classpath will be placed earlier in the interceptor order than
+ *     interceptors later in the classpath. Interceptors earlier within a specific file on the classpath will be placed earlier in
+ *     the order than interceptors later in the file.</li>
+ *
+ *     <li><i>Override Configuration Interceptors</i>. Any interceptors registered using
+ *     {@link software.amazon.awssdk.config.ClientOverrideConfiguration.Builder#addLastExecutionInterceptor(ExecutionInterceptor)}
+ *     in the order they were added.</li>
+ * </ol>
+ * When a request is being processed (up to and including {@link #beforeTransmission}, interceptors are applied in forward-order,
+ * according to the order described above. When a response is being processed (after and including {@link #afterTransmission},
+ * interceptors are applied in reverse-order from the order described above. This means that the last interceptors to touch the
+ * request are the first interceptors to touch the response.
+ * </p>
+ */
+public interface ExecutionInterceptor {
+    /**
+     * Read a request that has been given to a service client before it is modified by other interceptors.
+     * {@link #beforeMarshalling} should be used in most circumstances for reading the request because it includes modifications
+     * made by other interceptors.
+     *
+     * This method is guaranteed to be executed on the thread that is making the service call. This is true even if a non-blocking
+     * I/O client is used. This is useful for transferring data that may be stored thread-locally into the execution's
+     * {@link ExecutionAttributes}.
+     *
+     * @param execution The current state of the execution, including the unmodified SDK request from the service client call.
+     * @param executionAttributes A mutable set of attributes scoped to one specific request/response cycle that can be used to
+     *                            give data to future lifecycle methods.
+     */
+    default void beforeExecution(BeforeExecutionContext execution, ExecutionAttributes executionAttributes) {
+
+    }
+
+    /**
+     * Modify an {@link SdkRequest} given to a service client before it is marshalled into an {@link SdkHttpFullRequest}.
+     *
+     * @param execution The current state of the execution, including the current SDK request from the service client call.
+     * @param executionAttributes A mutable set of attributes scoped to one specific request/response cycle that can be used to
+     *                            give data to future lifecycle methods.
+     * @return The potentially-modified request that should be used for the rest of the execution. Must not be null.
+     */
+    default SdkRequest modifyRequest(BeforeMarshallingContext execution, ExecutionAttributes executionAttributes) {
+        return execution.request();
```

Comment:
That's fair. Would it be less confusing if the modify* methods had yet another context?

```Java
interface ExecutionInterceptor {
    SdkRequest modifyRequest(ModifyRequestContext execution, ExecutionAttributes executionAttributes);
}
```